### PR TITLE
support number for logging with sync_dist=True

### LINKED ISSUE #5003: AttributeError: 'float' object has no attribute 'clone' - when logging float values with sync_dist=True and ddp
## 🐛 Bug
```
self.log('d_acc_dummy' , 100.0, on_step=True, on_epoch=True, sync_dist=True) 
```
results in 
```
/usr/local/lib/python3.6/dist-packages/pytorch_lightning/core/step_result.py in log(self, name, value, prog_bar, logger, on_step, on_epoch, reduce_fx, tbptt_reduce_fx, tbptt_pad_token, enable_graph, sync_dist, sync_dist_op, sync_dist_group, sync_fn)
    138             is_dist_initialized = torch.distributed.is_available() and torch.distributed.is_initialized()
    139             # TODO: Find a way to make the reduction only once, so we don't need to clone.
--> 140             value = value.clone() if is_dist_initialized else value
    141             value = sync_fn(value, group=sync_dist_group, reduce_op=sync_dist_op)
    142 

AttributeError: 'float' object has no attribute 'clone'
```
Float is allowed type (from step_result.py)
```
 if sync_dist and isinstance(value, (torch.Tensor, numbers.Number)):
            is_dist_initialized = torch.distributed.is_available() and torch.distributed.is_initialized()
            # TODO: Find a way to make the reduction only once, so we don't need to clone.
            value = value.clone() if is_dist_initialized else value
            value = sync_fn(value, group=sync_dist_group, reduce_op=sync_dist_op)
```

## Please reproduce using [the BoringModel and post here]
[Simple gan example](https://colab.research.google.com/drive/1_3ZyOfmvdYl5zGEaF-WBcYd1aG0w4k9y?usp=sharing)

### Expected behavior
Logging of float values

### Environment
* CUDA:
	- GPU:
		- Tesla T4
	- available:         True
	- version:           10.1
* Packages:
	- numpy:             1.18.5
	- pyTorch_debug:     True
	- pyTorch_version:   1.7.0+cu101
	- pytorch-lightning: 1.0.8
	- tqdm:              4.41.1
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- 
	- processor:         x86_64
	- python:            3.6.9
	- version:           #1 SMP Thu Jul 23 08:00:38 PDT 2020

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR clone value only if it is a tensor.

Fixes #5003

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:  

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified; _Bugfixes should be including in bug-fix release milestones (m.f.X) and features should be included in (m.X.b) releases._
 

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/core/step_result.py [hunk 1]
```diff
@@ -128,6 +128,7 @@ def log(
         sync_dist_group: Optional[Any] = None,
         sync_fn: Callable = None,
         dataloader_idx: Optional[int] = None,
+        device: torch.device = None,
```
Comment (Borda) [inline pytorch_lightning/core/step_result.py:131]: let's add an explanation to the docs

File: tests/trainer/logging_tests/test_train_loop_logging_1_0.py [hunk 2]
```diff
@@ -685,6 +686,7 @@ class TestModel(BoringModel):
         def training_step(self, batch, batch_idx):
             acc = self.step(batch[0])
             self.log('foo', torch.tensor(fake_result), on_step=False, on_epoch=True, sync_dist=True, sync_dist_op='sum')
+            self.log('foo_2', 2, on_step=False, on_epoch=True, sync_dist=True, sync_dist_op='sum')
```
Comment (Borda) [inline tests/trainer/logging_tests/test_train_loop_logging_1_0.py:689]: ❤️

File: tests/trainer/logging_tests/test_train_loop_logging_1_0.py [hunk 3]
```diff
@@ -704,9 +706,46 @@ def validation_step(self, batch, batch_idx):
...
 
     assert trainer.logged_metrics['foo'] == fake_result
+    assert trainer.logged_metrics['foo_2'] == 2
     assert trainer.logged_metrics['bar'] == fake_result
 
 
+@pytest.mark.skipif(torch.cuda.device_count() < 2, reason="test requires multi-GPU machine")
+@pytest.mark.skipif(not os.getenv("PL_RUNNING_SPECIAL_TESTS", '0') == '1',
+                    reason="test should be run outside of pytest")
+def test_logging_sync_dist_true_ddp(tmpdir):
```
Comment (Borda) [inline tests/trainer/logging_tests/test_train_loop_logging_1_0.py:716]: add it to the DDP list?

File: pytorch_lightning/core/step_result.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/core/step_result.py:131]: Not sure I understand this comment. Result.log is an internal function used by LightningModule.log. It shouldn't be used by a user.

File: tests/trainer/logging_tests/test_train_loop_logging_1_0.py (hunk 3 above)
Comment (tchaton) [reply tests/trainer/logging_tests/test_train_loop_logging_1_0.py:716]: It has been added already.
```
python ${DEFAULTS} tests/trainer/logging_tests/test_train_loop_logging_1_0.py::test_logging_sync_dist_true_ddp
```

**Review by Borda — APPROVED** [review]
lgtm

**Review by justusschock — APPROVED** [review]
Lgtm

File: pytorch_lightning/core/step_result.py [hunk 4]
```diff
@@ -138,7 +139,10 @@ def log(
         if sync_dist and isinstance(value, (torch.Tensor, numbers.Number)):
             is_dist_initialized = torch.distributed.is_available() and torch.distributed.is_initialized()
             # TODO: Find a way to make the reduction only once, so we don't need to clone.
-            value = value.clone() if is_dist_initialized else value
+            if is_dist_initialized and isinstance(value, torch.Tensor):
+                value = value.clone()
+            else:
+                value = torch.tensor(value, device=device).float()
```
Comment (carmocca) [inline pytorch_lightning/core/step_result.py:145]:
```suggestion
                value = torch.tensor(value, device=device, dtype=torch.float)
```

File: pytorch_lightning/utilities/distributed.py [hunk 5]
```diff
@@ -145,9 +145,9 @@ def sync_ddp(
     if group is None:
         group = torch.distributed.group.WORLD
 
-    if reduce_op is None:
+    if reduce_op is None or (isinstance(reduce_op, str) and reduce_op.lower() == "sum"):
         reduce_op = torch.distributed.ReduceOp.SUM
-    elif isinstance(reduce_op, str) and reduce_op in ("avg", "mean"):
+    elif isinstance(reduce_op, str) and reduce_op.lower() in ("avg", "mean"):
         reduce_op = torch.distributed.ReduceOp.SUM
```
Comment (carmocca) [inline pytorch_lightning/utilities/distributed.py:151]: You didn't write this, but why is `ReduceOp.SUM` used for both cases? It should be outside of the `if-elif`

Comment (tchaton) [reply pytorch_lightning/utilities/distributed.py:151]: Good catch ! It can be simplified. We just need to divide by world size.

**Review by rohitgr7 — APPROVED** [review]

**Review by carmocca — APPROVED** [review]

**Review by SkafteNicki — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]
it is very strange that all other tests that use sync_dist=True only run on a single gpu. these tests should probably be revised

**Comment by tchaton** [comment]
> it is very strange that all other tests that use sync_dist=True only run on a single gpu. these tests should probably be revised

Yes, entirely agree. I was confused too when I looked at them. We definitely need better tests around sync.

Best,

**Pull request merged**